pennylane/templates/embeddings/amplitude.py

```diff
@@ -15,27 +15,128 @@
 Contains the ``AmplitudeEmbedding`` template.
 """
 # pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
-import math
+import warnings
 import numpy as np
+
+import pennylane as qml
 from pennylane.templates.decorator import template
 from pennylane.ops import QubitStateVector
-from pennylane.templates.utils import (
-    check_shape,
-    check_type,
-    get_shape,
-)
 from pennylane.variable import Variable
 from pennylane.wires import Wires
+from pennylane.templates.utils import check_shape, get_shape, check_type
 
 # tolerance for normalization
 TOLERANCE = 1e-10
 
 
+def _preprocess(features, wires, pad_with, normalize):
+    """Validate and pre-process inputs."""
+
+    if qml.tape_mode_active():
+
+        features = qml.tensorbox.TensorBox(features)
+
+        # check shape
+        if len(features.shape) != 1:
+            raise ValueError(f"Features must be a one-dimensional vector; got shape {features.shape}.")
+
+        n_features = features.shape[0]
+        if pad_with is None and n_features != 2 ** len(wires):
+            raise ValueError(f"Features must be of length {2 ** len(wires)}; got length {n_features}. "
+                             f"Use the 'pad' argument for automated padding.")
+
+        if pad_with is not None and n_features > 2 ** len(wires):
+            raise ValueError(f"Features must be of length {2 ** len(wires)} or "
+                             f"smaller to be padded; got length {n_features}.")
+
+        # pad
+        if pad_with is not None and n_features < 2 ** len(wires):
+            padding = [pad_with] * (2 ** len(wires) - n_features)
+            features = qml.tensorbox.concatenate([features, padding], axis=0)
+
+        #TODO: add those methods to tensorbox
+
+        # normalize
+        norm = np.sum(np.abs(features) ** 2)
+
+        if not np.isclose(norm, 1.0, atol=TOLERANCE):
```

Comment:
This needs to be refactored to agnostic processing.

Reply:
Since we just want to extract the scalar norm, one could also just extract a numpy representation of the input here (for feature parity at least, I think this stops us from being diffable). But I had trouble doing that, `features.numpy()` did not work in all cases for some reason.